nowriter/tmencoding/tm_encoder_bytesout_builder_adaptor.go

```diff
@@ -0,0 +1,94 @@
+package tmencoding
+
+import "bytes"
+import "time"
+
+type TMEncoderBytesOutBuilderAdaptor struct {
+	TMEncoderBuilderIntr
+	buf  bytes.Buffer
+	pure TMEncoderBytesOutIntr
+}
+
+func (a *TMEncoderBytesOutBuilderAdaptor) Bytes() []byte {
+	return a.buf.Bytes()
+}
+
+func (a *TMEncoderBytesOutBuilderAdaptor) WriteBool(b bool) TMEncoderBuilderIntr {
+	a.buf.Write(a.pure.WriteBool(b))
```

Comment:
Okay, this is actually a design comment.

bytes.Buffer has an internal byte slice, when you append to it, you do a copy, and possibly allocate memory (I think it keeps allocating twice the space and many appends just copy until it hits the end of the space, when it must malloc). It will not modify the data passed in.

pure allocates a byte slice just to be appended, and then released. for lots of encoding, this increases gc pressure with plenty of small slices.

this same problem exists in the legacy code, so not a critique of your code.

just a thought if it would be possible to make an implementation that didn't make those extra allocations. hmmm... actually if all code refers only to `TMEncoderBuilderIntr`, then I am sure that one could write a `TMEncoderBytesOutOptimizedBuilderAdaptor` that was just some hyper-optimized version of this class and it could be swapped in later.

Reply:
i wanted to focus on clarity at the exclusion of performance discussion for this pr. can we move performance discussion to another pr or github issue please after we have benchmarking to justify using dev time on it?

Reply:
BTW i think C++ solved this issue in 2011 with the rvalue semantics (moveTo) and Rust is built around the solution to this problem with the move semantic. @ethanfrey

Reply:
You forget that Go was actually designed in the 70s :grin:

Reply:
intentional based on @jaekwon dictate and also simpler at only minor performance cost at worst. in practice much simpler to develop using pure function style with no side effects / state in large part.